contrib/JLightning/src/de/renepickhardt/ln/UnixDomainSocketRpc.java

```diff
@@ -0,0 +1,119 @@
+package de.renepickhardt.ln;
+/**
+ * UnixDomainSocketRpc the base class to handle communication between
+ * JLightning and the c-lightning node over the UnixDomainSocket 
+ * 
+ * This file is basically a port of the pylightning python client library
+ * that comes with c-lightning.
+ * 
+ * The Author of this Java Client library is Rene Pickhardt. 
+ * He also holds the copyright of this file. The library is licensed with
+ * a BSD-style license. Have a look at the LICENSE file. 
+ * 
+ * If you like this library consider a donation via bitcoin or the lightning
+ * network at http://ln.rene-pickhardt.de
+ * 
+ * @author Rene Pickhardt
+ */
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.newsclub.net.unix.AFUNIXSocket;
+import org.newsclub.net.unix.AFUNIXSocketAddress;
+
+public class UnixDomainSocketRpc {
+	protected AFUNIXSocket sock;
+	protected InputStream is;
+	protected OutputStream os;
+	protected static int id = 1;
+	
+	public UnixDomainSocketRpc(String socket_path) {
+		File socketFile = new File(socket_path);
+		try {
+			this.sock = AFUNIXSocket.newInstance();
+			this.sock.connect(new AFUNIXSocketAddress(socketFile));
+			this.is = sock.getInputStream();
+			this.os = sock.getOutputStream();			
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+	}
+	
+	public String call(String method, HashMap<String, String> payload) {
+		// if no payload is given make empty one
+		if(payload == null) {
+			payload = new HashMap<String, String>();
+		}
+		
+		//remove null items from payload
+		Set<String> keySet = new HashSet<String>();
+		for (String key: payload.keySet()) {
+			keySet.add(key);
+		}
+		for(String k:keySet) {
+			if(payload.get(k)==null){
+				payload.remove(k);
+			}
+		}
+		
+		JSONObject json = new JSONObject();
+		
+		try {
+			json.put("method", method);
+			json.put("params", new JSONObject(payload));
+			// FIXME: Use id field to dispatch requests
+			json.put("id", Integer.toString(this.id++));
+		} catch (JSONException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
+		}
+		
+		try {
+			this.os.write(json.toString().getBytes("UTF-8"));
+			this.os.flush();
+			
+			// FIXME: Using a StringBuilder would be preferable but not so easy
+			String response = "";
+			int buffSize = 1024;
+			byte[] buf = new byte[buffSize];
+
+			while(true) { 
+				int read = this.is.read(buf);
+				response = response + new String(buf, 0, read, "UTF-8");
+				if(read <=0 || response.contains("\n\n" )) {
```

Comment:
in my Haskell implementation I count open and closed curly brackets (ensuring they sum to 0) to determine when to stop reading.

This could fail in an unlikely event that both `\n`'s appear at the end and start of two separate reads. This would result in the entire program blocking indefinitely on the next read call.

Reply:
Response is the string that has been appended with the buffer so I see no problem with split \n

On the other side the sum of open and closed brakets might fail as those could be part of the strings in the json data. So those need to be ignored